src/reliability.nim

```diff
@@ -0,0 +1,339 @@
+
+      return serializeMessage(msg)
+    except Exception:
+      error "Failed to wrap message", msg = getCurrentExceptionMsg()
+      return err(ReliabilityError.reSerializationError)
+
+proc processIncomingBuffer(rm: ReliabilityManager) {.gcsafe.} =
+  withLock rm.lock:
+    if rm.incomingBuffer.len == 0:
+      return
+
+    # Create dependency map
+    var dependencies = initTable[SdsMessageID, seq[SdsMessageID]]()
+    var readyToProcess: seq[SdsMessageID] = @[]
+    var processed = initHashSet[SdsMessageID]()
+
+    # Build dependency graph and find initially ready messages
+    for msg in rm.incomingBuffer:
+      var hasMissingDeps = false
+      for depId in msg.causalHistory:
+        if not rm.bloomFilter.contains(depId):
+          hasMissingDeps = true
+          if depId notin dependencies:
+            dependencies[depId] = @[]
+          dependencies[depId].add(msg.messageId)
+
+      if not hasMissingDeps:
+        readyToProcess.add(msg.messageId)
+
+    while readyToProcess.len > 0:
+      let msgId = readyToProcess.pop()
+      if msgId in processed:
+        continue
+
+      # Process this message
+      for msg in rm.incomingBuffer:
+        if msg.messageId == msgId:
+          rm.addToHistory(msg.messageId)
+          if not rm.onMessageReady.isNil():
+            rm.onMessageReady(msg.messageId)
+          processed.incl(msgId)
+
+          # Add any dependent messages that might now be ready
+          if msgId in dependencies:
+            readyToProcess.add(dependencies[msgId])
+          break
+
+    rm.incomingBuffer = rm.incomingBuffer.filterIt(it.messageId notin processed)
+
+proc unwrapReceivedMessage*(
+    rm: ReliabilityManager, message: seq[byte]
+): Result[tuple[message: seq[byte], missingDeps: seq[SdsMessageID]], ReliabilityError] =
+  ## Unwraps a received message and processes its reliability metadata.
+  ##
+  ## Parameters:
+  ##   - message: The received message bytes
+  ##
+  ## Returns:
+  ##   A Result containing either tuple of (processed message, missing dependencies) or an error.
+  try:
+    let msg = deserializeMessage(message).valueOr:
+      return err(ReliabilityError.reDeserializationError)
+
+    if rm.bloomFilter.contains(msg.messageId):
+      return ok((msg.content, @[]))
+
+    rm.bloomFilter.add(msg.messageId)
+
+    # Update Lamport timestamp
+    rm.updateLamportTimestamp(msg.lamportTimestamp)
+
+    # Review ACK status for outgoing messages
+    rm.reviewAckStatus(msg)
+
+    var missingDeps: seq[SdsMessageID] = @[]
+    for depId in msg.causalHistory:
+      if not rm.bloomFilter.contains(depId):
+        missingDeps.add(depId)
+
+    if missingDeps.len == 0:
+      # Check if any dependencies are still in incoming buffer
+      var depsInBuffer = false
+      for bufferedMsg in rm.incomingBuffer:
+        if bufferedMsg.messageId in msg.causalHistory:
+          depsInBuffer = true
+          break
+
+      if depsInBuffer:
+        rm.incomingBuffer.add(msg)
```

Comment:
I don't think this step is necessary if we're going to add the message to the `incomingBuffer` in any case. Presumably, you'd like to avoid the `onMissingDependencies` call? What is expected to happen within the `onMissingDependencies` call?